syntax = "proto3";

message AABBf {
   float3 min = 1;
   float3 max = 2;
}
message AlbedoChannel {
   float4 Color = 1;
   string Texture = 2;
   float Mix = 3;
   oneof subtype {
      SpecularChannel SpecularChannel = 100;
   }
}
message BRDFChannel {
   float IOR = 1;
   float Metallic = 2;
   float Specular = 3;
   float Roughness = 4;
   float Subsurface = 5;
   float3 SubsurfaceColor = 6;
}
enum Distribution {
   GGX = 0;
}
message FusAnimation {
   repeated FusAnimationChannel AnimationChannel = 3;
}
message FusAnimationChannel {
   int32 SceneComponent = 1;
   string Property = 2;
   TypeId TypeId = 3;
   repeated FusAnimationKeyBase KeyFrames = 4;
   LerpType LerpType = 5;
}
message FusAnimationKeyBase {
   float Time = 1;
   oneof subtype {
      FusAnimationKeyDouble FusAnimationKeyDouble = 100;
      FusAnimationKeyInt FusAnimationKeyInt = 101;
      FusAnimationKeyFloat FusAnimationKeyFloat = 102;
      FusAnimationKeyFloat2 FusAnimationKeyFloat2 = 103;
      FusAnimationKeyFloat3 FusAnimationKeyFloat3 = 104;
      FusAnimationKeyFloat4 FusAnimationKeyFloat4 = 105;
   }
}
message FusAnimationKeyDouble {
   double Value = 1;
}
message FusAnimationKeyFloat {
   float Value = 1;
}
message FusAnimationKeyFloat2 {
   float2 Value = 1;
}
message FusAnimationKeyFloat3 {
   float3 Value = 1;
}
message FusAnimationKeyFloat4 {
   float4 Value = 1;
}
message FusAnimationKeyInt {
   int32 Value = 1;
}
message FusBone {
   string Name = 1;
}
message FusCamera {
   bool ClearColor = 1;
   bool ClearDepth = 2;
   int32 Layer = 3;
   float4 BackgroundColor = 4;
   ProjectionMethod ProjectionMethod = 5;
   float Fov = 6;
   float2 ClippingPlanes = 7;
   float4 Viewport = 8;
   float Scale = 9;
}
message FusComponent {
   string Name = 1;
   bool Active = 2;
   oneof subtype {
      FusTransform FusTransform = 100;
      FusMesh FusMesh = 101;
      FusMaterialBase FusMaterialBase = 102;
      FusLight FusLight = 103;
      FusWeight FusWeight = 104;
      FusAnimation FusAnimation = 105;
      FusBone FusBone = 106;
      FusCamera FusCamera = 107;
   }
}
message FusContents {
   oneof subtype {
      FusScene FusScene = 201;
   }
}
message FusFile {
   FusHeader Header = 1;
   FusContents Contents = 2;
}
message FusHeader {
   int32 FileVersion = 1;
   string Generator = 2;
   string CreatedBy = 3;
   string CreationDate = 4;
}
message FusLight {
   float4 Color = 1;
   float MaxDistance = 2;
   float Strength = 3;
   LightType Type = 4;
   float OuterConeAngle = 5;
   float InnerConeAngle = 6;
}
message FusMaterialBRDF {
   BRDFChannel BRDF = 2;
}
message FusMaterialBase {
   AlbedoChannel Albedo = 1;
   AlbedoChannel Emissive = 3;
   NormapMapChannel NormalMap = 4;
   oneof subtype {
      FusMaterialStandard FusMaterialStandard = 200;
      FusMaterialBRDF FusMaterialBRDF = 201;
      FusMaterialDiffuseBRDF FusMaterialDiffuseBRDF = 202;
      FusMaterialGlossyBRDF FusMaterialGlossyBRDF = 203;
   }
}
message FusMaterialDiffuseBRDF {
   float Roughness = 1;
}
message FusMaterialGlossyBRDF {
   float Roughness = 1;
   Distribution Distribution = 2;
}
message FusMaterialStandard {
   AlbedoChannel Specular = 2;
}
message FusMesh {
   repeated float3 Vertices = 1;
   repeated uint32 Colors = 2 [packed = false];
   repeated float3 Normals = 3;
   repeated float2 UVs = 4;
   repeated float4 BoneWeights = 5;
   repeated float4 BoneIndices = 6;
   repeated uint32 Triangles = 7 [packed = false];
   AABBf BoundingBox = 8;
   repeated float4 Tangents = 9;
   repeated float3 BiTangents = 10;
   int32 MeshType = 11;
}
message FusNode {
   string Name = 1;
   repeated int32 Components = 2 [packed = false];
   repeated FusNode Children = 3;
}
message FusScene {
   repeated FusComponent ComponentList = 1;
   repeated FusNode Children = 2;
}
message FusTransform {
   float3 Translation = 1;
   float3 Rotation = 2;
   float3 Scale = 3;
}
message FusWeight {
   repeated VertexWeightList WeightMap = 1;
   repeated FusComponent Joints = 2;
   repeated float4x4 BindingMatrices = 3;
}
enum LerpType {
   Lerp = 0;
   Slerp = 1;
}
enum LightType {
   Point = 0;
   Parallel = 1;
   Spot = 2;
   Legacy = 3;
}
message NormapMapChannel {
   string Texture = 1;
   float Intensity = 2;
}
enum ProjectionMethod {
   Perspective = 0;
   Orthographic = 1;
}
message SpecularChannel {
   float Shininess = 1;
   float Strength = 2;
}
enum TypeId {
   Double = 0;
   Int = 1;
   Float = 2;
   Float2 = 3;
   Float3 = 4;
   Float4 = 5;
   Bool = 6;
}
message VertexWeight {
   int32 JointIndex = 1;
   float Weight = 2;
}
message VertexWeightList {
   repeated VertexWeight VertexWeights = 1;
}
message float2 {
   float x = 1;
   float y = 2;
}
message float3 {
   float x = 1;
   float y = 2;
   float z = 3;
}
message float4 {
   float x = 1;
   float y = 2;
   float z = 3;
   float w = 4;
}
message float4x4 {
   float4 Row1 = 1;
   float4 Row2 = 2;
   float4 Row3 = 3;
   float4 Row4 = 4;
}
